### ISSUE TITLE
nil dereference in policy.ruleSlice.updateEndpointsCaches

### ISSUE THREAD

**Issue by raybejjani** [cilium/cilium#7782]
I noticed this in a nightly failure. I'm looking at it (https://github.com/cilium/cilium/pull/7780) but reproducing it is a little difficult. My theory is that the nil is the Identity returned by GetSecurityIdentity.

There are two things:
1- The implementation on Endpoint assumes the lock is held, but the Endpoint interface in-scope with EventQueue doesn't have locks. I'm experimenting with making pkg/endpoint/Endpoint.GetSecurityIdentity take the lock.
2- We don't check that Identity is nil, and I think that just means the endpoint is not initialised yet, and thus Rule.matches should return nil (or updateEndpointsCaches can return early).

```
2019-04-14T00:29:20.734695714Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=678ns eventEnqueueWaitTime=773ns eventHandlingDuration="474.899µs" eventType="*main.PolicyDeleteEvent" name=repository-change-queue subsys=eventqueue
2019-04-14T00:29:20.738211783Z panic: runtime error: invalid memory address or nil pointer dereference
2019-04-14T00:29:20.738236307Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x231b390]
2019-04-14T00:29:20.738251405Z 
2019-04-14T00:29:20.738255152Z goroutine 28050 [running]:
2019-04-14T00:29:20.738258311Z github.com/cilium/cilium/pkg/policy.(*rule).matches(0xc0015e8690, 0x80c, 0x0, 0x1114200)
2019-04-14T00:29:20.73826143Z 	/go/src/github.com/cilium/cilium/pkg/policy/rule.go:469 +0xb0
2019-04-14T00:29:20.738264831Z github.com/cilium/cilium/pkg/policy.ruleSlice.updateEndpointsCaches(0xc00119ca30, 0x2, 0x2, 0x33ef980, 0xc001bc62c0, 0xc000a7d600, 0x0)
2019-04-14T00:29:20.738268813Z 	/go/src/github.com/cilium/cilium/pkg/policy/rules.go:284 +0xac
2019-04-14T00:29:20.738271967Z github.com/cilium/cilium/pkg/policy.ruleSlice.UpdateRulesEndpointsCaches.func1(0x33ef980, 0xc001bc62c0)
2019-04-14T00:29:20.738274979Z 	/go/src/github.com/cilium/cilium/pkg/policy/repository.go:462 +0x66
2019-04-14T00:29:20.738278027Z github.com/cilium/cilium/pkg/policy.(*EndpointSet).ForEach.func1(0xc00090dc80, 0xc000a2c0b0, 0x33ef980, 0xc001bc62c0)
2019-04-14T00:29:20.738280977Z 	/go/src/github.com/cilium/cilium/pkg/policy/identifier.go:75 +0x3a
2019-04-14T00:29:20.738284152Z created by github.com/cilium/cilium/pkg/policy.(*EndpointSet).ForEach
2019-04-14T00:29:20.738287776Z 	/go/src/github.com/cilium/cilium/pkg/policy/identifier.go:74 +0x12c
2019-04-14T00:29:20.758414296Z level=fatal msg="Agent pipe unexpectedly closed, shutting down" subsys=cilium-node-monitor
```

**Comment by ianvernon** [comment]
If the identity is nil, I think `matches` should return `false`. I have `nil` checks for Identity being nil elsewhere in related functions. You are correct in that the endpoint may be initializing, but it also may be in the process of getting deleted as well. What were the logs indicating what stage of the endpoint lifecycle for the endpoint which was consuming this identity? I'd be interested in seeing those.

**Comment by raybejjani** [comment]
TL;DR This is `unexpected end of JSON` it seems.

I'm not sure I know which endpoint this is (since the pointer in these functions is the `Endpoint` interface instance, and not the underlying concrete type). From what I can tell:
- The last eventqueue log lines (with a 0.5s gap to the previous ones) are
```
2019-04-14T00:29:20.242832929Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime="530.49µs" eventEnqueueWaitTime=430ns eventHandlingDuration=6.170839167s eventType="*endpoint.EndpointRegenerationEvent" name=endpoint-3766 subsys=eventqueue
2019-04-14T00:29:20.242897095Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=6.144692433s eventEnqueueWaitTime="1.777µs" eventHandlingDuration="26.783µs" eventType="*endpoint.EndpointRevisionBumpEvent" name=endpoint-3766 subsys=eventqueue
2019-04-14T00:29:20.242957606Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=6.140474082s eventEnqueueWaitTime="2.101µs" eventHandlingDuration="24.629µs" eventType="*endpoint.EndpointRevisionBumpEvent" name=endpoint-3766 subsys=eventqueue
2019-04-14T00:29:20.622624831Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime="403.941µs" eventEnqueueWaitTime=438ns eventHandlingDuration=6.51760994s eventType="*endpoint.EndpointRegenerationEvent" name=endpoint-2183 subsys=eventqueue
2019-04-14T00:29:20.622627247Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=6.517965372s eventEnqueueWaitTime="1.766µs" eventHandlingDuration="6.421µs" eventType="*endpoint.EndpointRevisionBumpEvent" name=endpoint-2183 subsys=eventqueue
2019-04-14T00:29:20.622629628Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=6.461391847s eventEnqueueWaitTime="5.988µs" eventHandlingDuration="4.479µs" eventType="*endpoint.EndpointRevisionBumpEvent" name=endpoint-2183 subsys=eventqueue
2019-04-14T00:29:20.734695714Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=678ns eventEnqueueWaitTime=773ns eventHandlingDuration="474.899µs" eventType="*main.PolicyDeleteEvent" name=repository-change-queue subsys=eventqueue
```

and these endpoints were restored much earlier.
- There seems to be a policy change happening:
```
2019-04-14T00:29:20.703881153Z level=debug msg="Queueing container event" containerID=4cdda2b70c event=stop subsys=workload-watcher
2019-04-14T00:29:20.703903819Z level=debug msg="Queueing container event" containerID=4cdda2b70c event=stop subsys=workload-watcher
2019-04-14T00:29:20.734257308Z level=debug msg="Deleting CiliumNetworkPolicy" ciliumNetworkPolicyName=3de13501 k8sApiVersion= k8sNamespace=default subsys=daemon
2019-04-14T00:29:20.734284268Z level=debug msg="Removed controller" name="sync-cnp-policy-status (v2 default/3de13501)" subsys=controller uuid=542e9dca-5e4c-11e9-badc-080027a49cd9
2019-04-14T00:29:20.734289219Z level=debug msg="Shutting down controller" name="sync-cnp-policy-status (v2 default/3de13501)" subsys=controller uuid=542e9dca-5e4c-11e9-badc-080027a49cd9
2019-04-14T00:29:20.734341351Z level=debug msg="Policy Delete Request" identityLabels="[k8s:io.cilium.k8s.policy.name=3de13501 k8s:io.cilium.k8s.policy.uid=244d3393-5e48-11e9-b682-080027a49cd9 k8s:io.cilium.k8s.policy.namespace=default k8s:io.cilium.k8s.policy.derived-from=CiliumNetworkPolicy]" subsys=daemon
2019-04-14T00:29:20.734606983Z level=debug msg="Policy deleted via API, found prefixes..." prefixes="[]" subsys=daemon
2019-04-14T00:29:20.734695714Z level=debug msg="EventQueue event processing statistics" eventConsumeOffQueueWaitTime=678ns eventEnqueueWaitTime=773ns eventHandlingDuration="474.899µs" eventType="*main.PolicyDeleteEvent" name=repository-change-queue subsys=eventqueue
2019-04-14T00:29:20.738211783Z panic: runtime error: invalid memory address or nil pointer dereference
```

so it could be that (which iterates over all endpoints and finds the bad one). Container `4cdda2b70c` seems to have a bunch die, stop and kill events from earlier, and I'm not sure how that can be.



The log file I'm looking at is 
[pod-kube-system-cilium-7rvtb-cilium-agent-previous.log](https://github.com/cilium/cilium/files/3107378/pod-kube-system-cilium-7rvtb-cilium-agent-previous.log) but I also have other logs. The pod list shows this cilium in a backoff, and one pod in ContainerCreating

```
NAMESPACE     NAME                                    READY   STATUS              RESTARTS   AGE   IP              NODE   NOMINATED NODE
default       0268ab2f-47d8b37b                       1/1     Running             0          28m   10.10.3.75      k8s4   <none>
default       0268ab2f-f07eed55                       1/1     Running             0          28m   10.10.2.7       k8s3   <none>
default       08e960ff-2557da1e                       1/1     Running             0          20m   10.10.2.121     k8s3   <none>
default       08e960ff-7f95b638                       1/1     Running             0          20m   10.10.3.157     k8s4   <none>
default       16006464-85ddcfc3                       1/1     Running             0          16m   10.10.0.54      k8s1   <none>
default       16006464-b5725134                       1/1     Running             0          16m   10.10.1.60      k8s2   <none>
default       172debb5-38d11495                       1/1     Running             0          28m   10.10.0.185     k8s1   <none>
default       172debb5-d9390532                       1/1     Running             0          28m   10.10.1.173     k8s2   <none>
default       1ad92c7c-2fa40c06                       1/1     Running             0          13m   10.10.0.77      k8s1   <none>
default       1ad92c7c-36b3032e                       1/1     Running             0          13m   10.10.1.68      k8s2   <none>
default       24f597cd-0376f2d7                       1/1     Running             0          24m   10.10.3.202     k8s4   <none>
default       24f597cd-6bacad83                       1/1     Running             0          24m   10.10.1.50      k8s2   <none>
default       30fba0f1-17c4a655                       1/1     Running             0          21m   10.10.1.175     k8s2   <none>
default       30fba0f1-4bd8039f                       1/1     Running             0          21m   10.10.0.147     k8s1   <none>
default       38617104-0d8a179f                       1/1     Running             0          29m   10.10.3.17      k8s4   <none>
default       38617104-4880f931                       1/1     Running             0          29m   10.10.2.210     k8s3   <none>
default       39ab0fc0-3ca8dcd0                       1/1     Running             0          19m   10.10.1.10      k8s2   <none>
default       39ab0fc0-8ae4029c                       1/1     Running             0          19m   10.10.0.218     k8s1   <none>
default       42a67bb9-746ca961                       1/1     Running             0          14m   10.10.0.50      k8s1   <none>
default       42a67bb9-92fbae7d                       1/1     Running             0          14m   10.10.1.205     k8s2   <none>
default       45a2dc0e-d121b44d                       0/1     ContainerCreating   0          11m   <none>          k8s2   <none>
default       45a2dc0e-e7664fbe                       1/1     Running             0          11m   10.10.0.198     k8s1   <none>
default       4ba62be6-42f1f842                       1/1     Running             0          17m   10.10.0.191     k8s1   <none>
default       4ba62be6-8c84699b                       1/1     Running             0          17m   10.10.1.251     k8s2   <none>
default       5d4bbe36-226f4fc2                       1/1     Running             0          18m   10.10.0.72      k8s1   <none>
default       5d4bbe36-a34d82e7                       1/1     Running             0          18m   10.10.1.195     k8s2   <none>
default       6162ba24-1bbcf90f                       1/1     Running             0          25m   10.10.1.176     k8s2   <none>
default       6162ba24-794afbd7                       1/1     Running             0          25m   10.10.0.109     k8s1   <none>
default       66e0b9f1-17cb08d0                       1/1     Running             0          20m   10.10.3.114     k8s4   <none>
default       66e0b9f1-5ce99b30                       1/1     Running             0          20m   10.10.2.35      k8s3   <none>
default       69c917ef-3f109210                       1/1     Running             0          23m   10.10.1.216     k8s2   <none>
default       69c917ef-6313baec                       1/1     Running             0          23m   10.10.0.62      k8s1   <none>
default       6d0b7728-d2b131c6                       1/1     Running             0          16m   10.10.2.123     k8s3   <none>
default       6d0b7728-f65a1585                       1/1     Running             0          16m   10.10.3.42      k8s4   <none>
default       6d2769fb-508a4b18                       1/1     Running             0          12m   10.10.2.126     k8s3   <none>
default       6d2769fb-9fb530fc                       1/1     Running             0          12m   10.10.3.197     k8s4   <none>
default       7aca7bdf-7b95abd5                       1/1     Running             0          21m   10.10.3.10      k8s4   <none>
default       7aca7bdf-d72c4eed                       1/1     Running             0          21m   10.10.2.171     k8s3   <none>
default       7e421f2f-a3c0f7c4                       1/1     Running             0          22m   10.10.0.16      k8s1   <none>
default       7e421f2f-e3a1ca90                       1/1     Running             0          22m   10.10.1.182     k8s2   <none>
default       7ec9134d-849be759                       1/1     Running             0          24m   10.10.1.95      k8s2   <none>
default       7ec9134d-be9888dd                       1/1     Running             0          24m   10.10.0.52      k8s1   <none>
default       7f2da9fd-164cf03f                       1/1     Running             0          26m   10.10.3.154     k8s4   <none>
default       7f2da9fd-de0dcbac                       1/1     Running             0          26m   10.10.2.124     k8s3   <none>
default       82a29bbe-0a8c2183                       1/1     Running             0          26m   10.10.1.99      k8s2   <none>
default       82a29bbe-9ddd54ad                       1/1     Running             0          26m   10.10.0.87      k8s1   <none>
default       895f2069-09cb2f83                       1/1     Running             0          29m   10.10.1.172     k8s2   <none>
default       895f2069-6c51c142                       1/1     Running             0          29m   10.10.0.254     k8s1   <none>
default       8a4d6e1b-468646ba                       1/1     Running             0          24m   10.10.1.180     k8s2   <none>
default       8a4d6e1b-8ddbb719                       1/1     Running             0          24m   10.10.0.110     k8s1   <none>
default       8e492f95-3b9603ce                       1/1     Running             0          28m   10.10.3.132     k8s4   <none>
default       8e492f95-8e42e2e0                       1/1     Running             0          28m   10.10.2.232     k8s3   <none>
default       9020efce-74d5ce74                       1/1     Running             0          20m   10.10.0.182     k8s1   <none>
default       9020efce-a64e8974                       1/1     Running             0          20m   10.10.1.252     k8s2   <none>
default       93d299d7-cd2ec977                       1/1     Running             0          13m   10.10.1.69      k8s2   <none>
default       93d299d7-d2e154f2                       1/1     Running             0          13m   10.10.0.236     k8s1   <none>
default       94165fcb-29738063                       1/1     Running             0          27m   10.10.0.71      k8s1   <none>
default       94165fcb-66048aa2                       1/1     Running             0          27m   10.10.1.124     k8s2   <none>
default       9a878104-42ac36fc                       1/1     Running             0          18m   10.10.2.173     k8s3   <none>
default       9a878104-c2727a9a                       1/1     Running             0          18m   10.10.3.18      k8s4   <none>
default       b07f1d81-527e257a                       1/1     Running             0          29m   10.10.0.135     k8s1   <none>
default       b07f1d81-d281d41d                       1/1     Running             0          29m   10.10.1.140     k8s2   <none>
default       b570c2c8-489e3297                       1/1     Running             0          22m   10.10.3.63      k8s4   <none>
default       b570c2c8-89f0875e                       1/1     Running             0          22m   10.10.2.201     k8s3   <none>
default       b7a143f3-707c8324                       1/1     Running             0          23m   10.10.2.204     k8s3   <none>
default       b7a143f3-b64d734c                       1/1     Running             0          23m   10.10.3.137     k8s4   <none>
default       bd29b55b-3abd43e8                       1/1     Running             0          14m   10.10.2.111     k8s3   <none>
default       bd29b55b-6da3f5f5                       1/1     Running             0          14m   10.10.3.198     k8s4   <none>
default       bd52d84b-3a266994                       1/1     Running             0          15m   10.10.2.242     k8s3   <none>
default       bd52d84b-ed11b2b2                       1/1     Running             0          15m   10.10.3.85      k8s4   <none>
default       bef33487-05457aa4                       1/1     Running             0          13m   10.10.3.238     k8s4   <none>
default       bef33487-ef28a6cb                       1/1     Running             0          13m   10.10.2.146     k8s3   <none>
default       d90f1e4f-78ff37a7                       1/1     Running             0          25m   10.10.3.89      k8s4   <none>
default       d90f1e4f-d4a733aa                       1/1     Running             0          25m   10.10.2.8       k8s3   <none>
default       db2c1ed3-080a73f5                       1/1     Running             0          27m   10.10.2.214     k8s3   <none>
default       db2c1ed3-77948192                       1/1     Running             0          27m   10.10.3.22      k8s4   <none>
default       e71183c4-286d1f71                       1/1     Running             0          19m   10.10.2.234     k8s3   <none>
default       e71183c4-90afe4a5                       1/1     Running             0          19m   10.10.3.67      k8s4   <none>
kube-system   cilium-49b59                            1/1     Running             0          52m   192.168.36.14   k8s4   <none>
kube-system   cilium-7rvtb                            0/1     CrashLoopBackOff    6          52m   192.168.36.12   k8s2   <none>
kube-system   cilium-etcd-4d9sr6bjpl                  1/1     Running             0          64m   10.10.2.182     k8s3   <none>
kube-system   cilium-etcd-mmr4bbxlmk                  1/1     Running             0          66m   10.10.1.242     k8s2   <none>
kube-system   cilium-etcd-operator-5798968997-kz2b9   1/1     Running             0          52m   192.168.36.12   k8s2   <none>
kube-system   cilium-etcd-qk4fjv4rrh                  1/1     Running             0          65m   10.10.3.180     k8s4   <none>
kube-system   cilium-l96tn                            1/1     Running             0          52m   192.168.36.11   k8s1   <none>
kube-system   cilium-operator-868d7445fb-6sz2m        1/1     Running             0          52m   10.10.3.104     k8s4   <none>
kube-system   cilium-sfm99                            1/1     Running             0          52m   192.168.36.13   k8s3   <none>
kube-system   coredns-779ffd89bd-ng2kx                1/1     Running             0          52m   10.10.3.199     k8s4   <none>
kube-system   etcd-k8s1                               1/1     Running             0          73m   192.168.36.11   k8s1   <none>
kube-system   etcd-operator-78bcbf4574-gttx7          1/1     Running             0          52m   10.10.3.11      k8s4   <none>
kube-system   kube-apiserver-k8s1                     1/1     Running             0          73m   192.168.36.11   k8s1   <none>
kube-system   kube-controller-manager-k8s1            1/1     Running             0          73m   192.168.36.11   k8s1   <none>
kube-system   kube-proxy-kdvzl                        1/1     Running             0          74m   192.168.36.11   k8s1   <none>
kube-system   kube-proxy-qqstn                        1/1     Running             0          69m   192.168.36.12   k8s2   <none>
kube-system   kube-proxy-x5vlt                        1/1     Running             0          67m   192.168.36.14   k8s4   <none>
kube-system   kube-proxy-xwmhq                        1/1     Running             0          68m   192.168.36.13   k8s3   <none>
kube-system   kube-scheduler-k8s1                     1/1     Running             0          73m   192.168.36.11   k8s1   <none>
```

I did find that pod in [kubelet-k8s2-1.12.log](https://github.com/cilium/cilium/files/3107415/kubelet-k8s2-1.12.log) and this seems to be the ` unexpected end of JSON` thing :/

```
Apr 14 00:19:49 k8s2 kubelet[4254]: I0414 00:19:49.623100    4254 reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "default-token-lr9tr" (UniqueName: "kubernetes.io/secret/04192aa2-5e4b-11e9-b682-080027a49cd9-default-token-lr9tr") pod "45a2dc0e-d121b44d" (UID: "04192aa2-5e4b-11e9-b682-080027a49cd9")
Apr 14 00:19:49 k8s2 kubelet[4254]: W0414 00:19:49.761321    4254 raw.go:87] Error while processing event ("/sys/fs/cgroup/cpu,cpuacct/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope": 0x40000100 == IN_CREATE|IN_ISDIR): inotify_add_watch /sys/fs/cgroup/cpu,cpuacct/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope: no such file or directory
Apr 14 00:19:49 k8s2 kubelet[4254]: W0414 00:19:49.761841    4254 raw.go:87] Error while processing event ("/sys/fs/cgroup/blkio/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope": 0x40000100 == IN_CREATE|IN_ISDIR): inotify_add_watch /sys/fs/cgroup/blkio/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope: no such file or directory
Apr 14 00:19:49 k8s2 kubelet[4254]: W0414 00:19:49.761983    4254 raw.go:87] Error while processing event ("/sys/fs/cgroup/memory/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope": 0x40000100 == IN_CREATE|IN_ISDIR): inotify_add_watch /sys/fs/cgroup/memory/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope: no such file or directory
Apr 14 00:19:49 k8s2 kubelet[4254]: W0414 00:19:49.762103    4254 raw.go:87] Error while processing event ("/sys/fs/cgroup/devices/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope": 0x40000100 == IN_CREATE|IN_ISDIR): inotify_add_watch /sys/fs/cgroup/devices/system.slice/run-rb1b5c620034c429b95edf5f424894160.scope: no such file or directory
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Processing CNI ADD request" args="&{b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b /proc/12455/ns/net eth0 IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=45a2dc0e-d121b44d;K8S_POD_INFRA_CONTAINER_ID=b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b /opt/cni/bin [123 34
99 110 105 86 101 114 115 105 111 110 34 58 34 34 44 34 110 97 109 101 34 58 34 99 105 108 105 117 109 34 44 34 116 121 112 101 34 58 34 99 105 108 105 117 109 45 99 110 105 34 125]}" eventUUID=047e1a6c-5e4b-11e9-aff4-080027a49cd9 subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Created veth pair" subsys=endpoint-connector vethPair="[tmpb5730 lxcb91f105b3961]"
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Configuring link" interface=eth0 ipAddr=10.10.1.118 netLink="&{LinkAttrs:{Index:154 MTU:1500 TxQLen:0 Name:eth0 HardwareAddr:fe:c5:e4:61:07:38 Flags:broadcast|multicast RawFlags:4098 ParentIndex:155 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc00034f4e8 Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil>
OperState:down NetNsID:0 NumTxQueues:0 NumRxQueues:0 Vfs:[]} PeerName:}" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Adding route" route="{Prefix:{IP:10.10.1.1 Mask:ffffffff} Nexthop:<nil> Local:<nil> Device: MTU:0 Proto:0 Scope:0 Table:0 Type:0}" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Adding route" route="{Prefix:{IP:0.0.0.0 Mask:00000000} Nexthop:10.10.1.1 Local:<nil> Device: MTU:1450 Proto:0 Scope:0 Table:0 Type:0}" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Configuring link" interface=eth0 ipAddr="f00d::a0a:100:0:97a5" netLink="&{LinkAttrs:{Index:154 MTU:1500 TxQLen:0 Name:eth0 HardwareAddr:fe:c5:e4:61:07:38 Flags:broadcast|multicast RawFlags:4098 ParentIndex:155 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc00034f4e8 Promisc:0 Xdp:<nil> EncapType:ether Prot
info:<nil> OperState:down NetNsID:0 NumTxQueues:0 NumRxQueues:0 Vfs:[]} PeerName:}" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Adding route" route="{Prefix:{IP:f00d::a0a:100:0:1 Mask:ffffffffffffffffffffffffffffffff} Nexthop:<nil> Local:<nil> Device: MTU:0 Proto:0 Scope:0 Table:0 Type:0}" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Adding route" route="{Prefix:{IP::: Mask:00000000000000000000000000000000} Nexthop:f00d::a0a:100:0:1 Local:<nil> Device: MTU:1450 Proto:0 Scope:0 Table:0 Type:0}" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=warning msg="Unable to create endpoint" containerID=b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b error="Put http:///var/run/cilium/cilium.sock/v1/endpoint/cilium-local:0: EOF" eventUUID=047e1a6c-5e4b-11e9-aff4-080027a49cd9 subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=warning msg="Unable to release IP" error="Delete http:///var/run/cilium/cilium.sock/v1/ipam/10.10.1.118: read unix @->/var/run/cilium/cilium.sock: read: connection reset by peer\nIs the agent running?" ipAddr=10.10.1.118 subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: level=warning msg="Unable to release IP" error="Delete http:///var/run/cilium/cilium.sock/v1/ipam/f00d::a0a:100:0:97a5: dial unix /var/run/cilium/cilium.sock: connect: connection refused\nIs the agent running?" ipAddr="f00d::a0a:100:0:97a5" subsys=cilium-cni
Apr 14 00:19:50 k8s2 kubelet[4254]: E0414 00:19:50.398895    4254 cni.go:310] Error adding network: unexpected end of JSON input
Apr 14 00:19:50 k8s2 kubelet[4254]: E0414 00:19:50.398954    4254 cni.go:278] Error while adding to cni network: unexpected end of JSON input
Apr 14 00:19:50 k8s2 kubelet[4254]: level=debug msg="Processing CNI DEL request" args="&{b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b /proc/12455/ns/net eth0 IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=45a2dc0e-d121b44d;K8S_POD_INFRA_CONTAINER_ID=b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b /opt/cni/bin [123 34
99 110 105 86 101 114 115 105 111 110 34 58 34 34 44 34 110 97 109 101 34 58 34 99 105 108 105 117 109 34 44 34 116 121 112 101 34 58 34 99 105 108 105 117 109 45 99 110 105 34 125]}" subsys=cilium-cni
Apr 14 00:20:02 k8s2 kubelet[4254]: level=warning msg="Errors encountered while deleting endpoint" error="[DELETE /endpoint/{id}][404] deleteEndpointIdNotFound " subsys=cilium-cni
Apr 14 00:20:02 k8s2 kubelet[4254]: E0414 00:20:02.686069    4254 remote_runtime.go:96] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = failed to set up sandbox container "b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b" network for pod "45a2dc0e-d121b44d": NetworkPlugin cni failed to set up pod "45a2dc0e-d121b44d_def
ault" network: unexpected end of JSON input
Apr 14 00:20:02 k8s2 kubelet[4254]: E0414 00:20:02.686115    4254 kuberuntime_sandbox.go:68] CreatePodSandbox for pod "45a2dc0e-d121b44d_default(04192aa2-5e4b-11e9-b682-080027a49cd9)" failed: rpc error: code = Unknown desc = failed to set up sandbox container "b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b" network for pod "45a2dc0e-d121b44d":
NetworkPlugin cni failed to set up pod "45a2dc0e-d121b44d_default" network: unexpected end of JSON input
Apr 14 00:20:02 k8s2 kubelet[4254]: E0414 00:20:02.686125    4254 kuberuntime_manager.go:657] createPodSandbox for pod "45a2dc0e-d121b44d_default(04192aa2-5e4b-11e9-b682-080027a49cd9)" failed: rpc error: code = Unknown desc = failed to set up sandbox container "b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b" network for pod "45a2dc0e-d121b44d":
 NetworkPlugin cni failed to set up pod "45a2dc0e-d121b44d_default" network: unexpected end of JSON input
Apr 14 00:20:02 k8s2 kubelet[4254]: E0414 00:20:02.686168    4254 pod_workers.go:186] Error syncing pod 04192aa2-5e4b-11e9-b682-080027a49cd9 ("45a2dc0e-d121b44d_default(04192aa2-5e4b-11e9-b682-080027a49cd9)"), skipping: failed to "CreatePodSandbox" for "45a2dc0e-d121b44d_default(04192aa2-5e4b-11e9-b682-080027a49cd9)" with CreatePodSandboxError: "CreatePodSandbox
 for pod \"45a2dc0e-d121b44d_default(04192aa2-5e4b-11e9-b682-080027a49cd9)\" failed: rpc error: code = Unknown desc = failed to set up sandbox container \"b5730618806be9f6aaf18e845ed340d76c1300caf14c34c71499a0f593bb669b\" network for pod \"45a2dc0e-d121b44d\": NetworkPlugin cni failed to set up pod \"45a2dc0e-d121b44d_default\" network: unexpected end of JSON in
put"
...
repeat DELETE request until the end of the log
```

**Comment by raybejjani** [comment]
chatting with aanm the JSON error itself might be a red herring. If so, then the agent was breakage looping and that caused all the other things to go wrong. The endpoint failed in CNI ADD and so it could have been permanently lacking an identity.

**Comment by raybejjani** [comment]
superseded by a better fix
